**Юрий, привет!**

Меня зовут Артемьева Саша, я буду проверять твой проект. Приятно познакомиться, хоть и заочно:)
    
Предлагаю общаться на «ты» :) Но если это не удобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Так выделены небольшие замечания. Одно-два таких замечания в проекте допустимы, но если их будет больше - тебе будет необходимо внести исправления. Это как тестовое задание при приеме на работу: очень много мелких ошибок могут стать причиной отказа кандидату. 

</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 


<div class="alert alert-info"> <b>Комментарий студента:</b> Саша, привет! Спасибо за "щадящие" и дипломатичные комментарии :)</div>




# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Образование — 
Семейное положение
Тип кредита — одобрение 

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
# Импортируем Pandas и сохраняем датасет в переменную df:

import pandas as pd
df = pd.read_csv('/datasets/data.csv')

In [2]:
data = pd.read_csv('/datasets/data.csv')
data.to_csv('data.csv', index=False)

In [3]:
# Посмотрим размер таблицы:

df.shape

(21525, 12)

In [4]:
# Какие столбцы есть в таблице:

df.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

In [5]:
#to_csv(): df.to_csv('data.csv')

In [6]:
# Посмотрим общуя информацию о таблице:

df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 11.3 MB


Названия столбцов совпадают с названиями в ТЗ, всё нормально

Сразу обращаем внимание, что в колонках _days_employed_ и _total_income_ большое количество пропусков (в каждом случае 2174 пропущенных значений)

In [7]:
# Напечатаем первые 10 строк таблицы:

df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [8]:
# Напечатаем последние 10 строк:

df.tail(10)

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

По выведенным данным делаем первые наблюдения:
* В колонке _days_employed_ (общий трудовой стаж в днях) стоят отрицательные значения
* В столбце _education_ (образование) значения написаны в разных регистрах




In [9]:
# Посмотрим уникальные значения в графах education, family_status, income_type:

df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

In [10]:
# сколько значений есть в таблице, с учетом неявных дубликатов:

df['education'].value_counts().count()

15

Здесь придется поработать с неявными дубликатами — при фактически пяти значениях в столбце из-за разных регистров наблюдаем 15 (!) значений. В дальнейшем наведем порядок и уберем дубликаты.

Посмотрим на столбws `'family_status'` и `'income_type'`

In [11]:
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [12]:
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

В этих столбцах не видно неявных дубликатов, с ними можно работать

Сделаем то же самое для столбцов _education_id_, _family_status_id_

In [13]:
df['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

In [14]:
df['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

Проверим цели для получения денег — посмотрим столбец `'purpose'`

In [15]:
df['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

In [16]:
# Посмотрим количество значений в этом столбце:

df['purpose'].value_counts().count()

38

Многие значения по сути одинаковы, например, "сделка с подержанным автомобилем" и "на покупку подержанного автомобиля".

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Увидели, с чем предстоит иметь дело

## Шаг 2. Предобработка данных

### Обработка пропусков

Посмотрим общее количество пропусков:

In [17]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

#### Обработка пропусков в days_employed

 В столбце `days_employed` три проблемы:
1. 2174 пропуска
2. отрицательные значения — в случае с рабочими днями явная ошибка
3. дробные числа — тоже ошибочно в случае с рабочими днями

In [18]:
# Уберем знак "-" из столбца days_employed:

df['days_employed'] = abs(df['days_employed'])

# сразу же проверим:

df.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [19]:
# Заменим пустые значения на медианные. Медианы возьмем разные для каждой группы занятости:

df['days_employed'] = df['days_employed'].fillna(df.groupby('income_type')['days_employed'].transform('median'))

# проверим отсутствие пропусков:

df.isna().sum()

children               0
days_employed          0
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [20]:
# убираем нецелочисленные значения из days_employed:

df['days_employed'] = df['days_employed'].astype('int')

# и сразу проверяем:

df.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   СРЕДНЕЕ             1   
8         2           6929         35    ВЫСШЕЕ             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [21]:
# посчитаем среднее арифметическое по days_employed:

df['days_employed'].mean().astype('int')

67299

In [22]:
# и посмотрим медиану:

df['days_employed'].median().astype('int')

1993

Значение 67299 дня = 184 года, при современном уровне медицины труднодостижимое значение трудового стажа. Значит, в столбце есть артефакты с ошибочно-большим значением. Проверим max значение в столбце

In [23]:
df['days_employed'].max()

401755

Получили максимум в 401755 дня, это 1 103 года. Возможно, эти значения ошибочно занесены в **часах**, тогда 401755 часов преобразится во вполне разумные 45 лет. Посмотрим на первые 10 максимальных значений в столбце:

In [24]:
df.sort_values(by='days_employed', ascending=False).head(10)

children  days_employed  dob_years education  education_id  \
6954          0         401755         56   среднее             1   
10006         0         401715         69    высшее             0   
7664          1         401675         61   среднее             1   
2156          0         401674         60   среднее             1   
7794          0         401663         61   среднее             1   
4697          0         401635         56   среднее             1   
13420         0         401619         63   Среднее             1   
17823         0         401614         59   среднее             1   
10991         0         401591         56   среднее             1   
8369          0         401590         58   среднее             1   

               family_status  family_status_id gender income_type  debt  \
6954          вдовец / вдова                 2      F   пенсионер     0   
10006  Не женат / не замужем                 4      F   пенсионер     0   
7664         женат / замужем                 0      F   пенсионер     0   
2156         женат / замужем                 0      M   пенсионер     0   
7794        гражданский брак                 1      F   пенсионер     0   
4697         женат / замужем                 0      F   пенсионер     0   
13420       гражданский брак                 1      F   пенсионер     0   
17823        женат / замужем                 0      F   пенсионер     0   
10991              в разводе                 3      F   пенсионер     0   
8369         женат / замужем                 0      F   пенсионер     0   

        total_income                                purpose  
6954   176278.441171                           ремонт жилью  
10006   57390.256908                  получение образования  
7664   126214.519212                      операции с жильем  
2156   325395.724541                             автомобили  
7794    48286.441362                                свадьба  
4697    48242.322502                   покупка недвижимости  
13420   51449.788325                        сыграть свадьбу  
17823  152769.694536                покупка жилья для сдачи  
10991   39513.517543  получение дополнительного образования  
8369   175306.312902                            образование

In [25]:
# Кол-во со стажем более 12775 дней:

df.loc[df['days_employed'] > 12775, ['days_employed']].count()


days_employed    3918
dtype: int64

In [26]:
# Кол-во со стажем более 300000 дней:

df.loc[df['days_employed'] > 300000, ['days_employed']].count()


days_employed    3858
dtype: int64

**Если гипотеза о том, что данные приведены в часах, верна, то первый в списке человек при возрасте 56 имеет трудовой стаж 45 лет, т.е. работает официально с 11 лет. То же самое можно сказать про людей в строках с индексом 4697, 8369,10991. В пользу гипотезы о часах говорит то, что все перечисленные люди относятся к категории "пенсионер". Но даже если список делался в 2000 году, эти люди родились в 1944 году, вряд ли в 1955 году была необходимость отправлять детей на работу в возрасте 11 лет. Поэтому гипотезу об ошибочном представлении данных в часах принять тяжело. С другой стороны, если взять средний трудовой стаж при выходе на пенсию — примерно 35 лет (https://ria.ru/20131119/977986173.html), получается 12775 дней. У нас людей, перешедших этот порог 3858 (18,2%). Более того, людей с указанным стажем в таблице более 300000 дней такое же количество. В нашем ТЗ нет необходимости узнать связь между стажем и платежеспособностью, но из-за ошибочных данных по 18% данных точно есть необходимость разобраться, откуда идет такая ошибка. 
Технически переделать стаж можно либо в медианное значение, либо сделать функцию для перевода данных из часов в дни.**

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Молодец, что обращаешься к помощи сторонних источников для подтверждения своих суждений!

#### Обработка пропусков в total_income

In [27]:
# Пора разобраться с пропусками в поле total_income. Заполним пустые значения медианным в зависимости от типа занятости:

df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform('median'))



In [28]:
# Проверим пропуски:

df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [29]:
# Выведем среднее арифметическое для total_income:

df['total_income'].mean()

165225.32451367722

In [30]:
# Теперь медиана для того же столбца:

df['total_income'].median()

142594.39684740017

In [31]:
# Посмотрим на min значения столбца:

df['total_income'].min()

20667.26379327158

In [32]:
# Посмотрим на max значения столбца:

df['total_income'].max()

2265604.028722744

In [33]:
# Проверим столбец на отрицательные значения:

df.loc[df['total_income'] < 0, ['total_income']].count()

total_income    0
dtype: int64

In [34]:
# Единственное замечание — большое количество знаков после запятой, вряд ли доход считается в стотысячных долях копейки. 
# Приведем значения к типу int

df['total_income'] = df['total_income'].astype('int')

В столбце нет экстремально высоких/низких значений, нет отрицательных значений. Медиана отличается от среднего арифметического на 15%. Предполагаем, что данные в столбце подходят для дальнейших исследований.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Проведена качественная обработка пропусков

#### Обработка пропусков в children

Посмотрим на один из ключевых столбцов — количество детей

In [35]:
# выведем уникальные значения столбца

df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Сразу несколько проблем:
1. Экстремально большое значение у 76 человек — 20 детей в семье
2. Отрицательное количество детей у 47 человек

Предположительный источник первой проблемы — замена значения 2.0 на 20
Отрицательные значения могли появиться из-за формы записи, например, в файле указывали количество детей через дефис. 
Посмотрим на долю аномальных значений в столбце `children` в процентах


In [36]:
anomal_children = ((df.loc[df['children'] == 20, ['children']].count()) + (df.loc[df['children'] < 0, ['children']].count())) / df['children'].count()
anomal_children


children    0.005714
dtype: float64

Всего 0,005%, этой цифрой можно смело принебречь. С помощью логической индексации поочередно удаляем строки со значениями 20 и меньше нуля

In [37]:
df = df.loc[df['children'] != 20]
df = df.loc[df['children'] >= 0]

In [38]:
# Проверим  

df['children'].value_counts()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

**Столбец children готов к работе: данные без пропусков, без отрицательных значений, без экстремальных всплесков, целочисленные**

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b> Сейчас доля аномалий мала и ни на что не повлияет. В следующий раз старайся сразу высчитывать их долю и если она меньше 5%, смело избавляйся от нее. Возможно наше предположение о том, что 20==2, а -1==1 неверно

</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Поправил.</div>




<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Очень хорошо!

#### Обработка пропусков в family_status и family_status_id

In [39]:
# Пропусков в столбце family_status нет, посмотрим на значения:

df['family_status'].value_counts()

женат / замужем          12302
гражданский брак          4160
Не женат / не замужем     2799
в разводе                 1189
вдовец / вдова             952
Name: family_status, dtype: int64

In [40]:
df['family_status'].value_counts().sum()

21402

In [41]:
# Сделаем то же самое для family_status_id:

df['family_status_id'].value_counts()

0    12302
1     4160
4     2799
3     1189
2      952
Name: family_status_id, dtype: int64

In [42]:
df['family_status_id'].value_counts().sum()

21402

Столбец family_status готов к работе, пропусков нет, некорректных значений нет.                                               
В столбце-идентификаторы family_status_id тоже корретные значения, количество категорий совпадает с количеством значений в family_status_id.

Посмотрим на столбец `gender`, пропусков в нем нет, убедимся в отсутствии аномальных значений

In [43]:
df['gender'].value_counts()

F      14154
M       7247
XNA        1
Name: gender, dtype: int64

Значение XNA, возможно, предполагает 3-й пол, либо трансгендера. Доля XNA — 0,004% от общего количества, для нашего исседования столбец "пол" не ключевой. Смело удаляем строки с этим значением через условную индексацию

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Верно!

In [44]:
df.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

Проверим колонку `debt`

In [45]:
df['debt'].value_counts()

0    19670
1     1732
Name: debt, dtype: int64

Здесь все хорошо, только 0 и 1 — были ли просроченные платежи или нет.

Проверим столбец `dob_years`

In [46]:
df['dob_years'].value_counts()

35    614
40    603
41    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
29    543
44    543
30    536
48    536
37    531
43    510
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
57    457
53    457
51    446
55    441
59    441
26    406
60    376
25    356
61    353
62    351
63    268
64    263
24    263
23    252
65    194
66    183
22    183
67    167
21    110
0     100
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Всё хорошо, кроме людей в возрасте 0 лет. Проверим % таких людей

In [47]:
(df.loc[df['dob_years'] == 0, 'dob_years'].count()) / df['dob_years'].count() * 100

0.4672460517708626

Менее 0,5%, смело удаляем эти данные через логическую индексацию

In [48]:
df = df.loc[df['dob_years'] != 0]
df['dob_years'].value_counts()

35    614
40    603
41    603
34    597
38    595
42    592
33    577
39    572
31    556
36    553
29    543
44    543
30    536
48    536
37    531
43    510
50    509
32    506
49    505
28    501
45    494
27    490
52    483
56    482
47    480
54    476
46    469
58    461
53    457
57    457
51    446
55    441
59    441
26    406
60    376
25    356
61    353
62    351
63    268
64    263
24    263
23    252
65    194
66    183
22    183
67    167
21    110
68     99
69     83
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Некорректные значения удалены, столбец готов к работе

**Вывод**

Ключевые столбцы children и family_status готовы к дальнейшим исследованиям.

В остальных столбцах убраны пропуски и значения нормализованы, аномальные значения удалены из-за своей ничтожно малой доли

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Что еще можно сказать об аномальных значениях? Они могут скрываться в любой колонке датафрейма, поэтому важно изучить каждый столбец </div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Поправил. В 'gender' убрал XNA, в 'dob_years' — 0</div>




<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Избавились от всех аномалий!

### Замена типа данных

**В пункте 2.1.1 меняли значение days_employed с float на int**

### Обработка дубликатов

#### Убираем неявные дубликаты

Посмотрим на неявные дубликаты. Они "поселились" в столбце education из-за набора одних и тех же значений в разных регистрах

In [49]:
# Убираем дубликаты методом str.lower():

df['education'] = df['education'].str.lower()

In [50]:
# Проверяем результат:

df['education'].value_counts()

среднее                15073
высшее                  5202
неоконченное высшее      739
начальное                282
ученая степень             6
Name: education, dtype: int64

Теперь в столбце 'education' 5 значений, как и должно быть

#### Убираем явные дубликаты

In [51]:
# Находим явные дубликаты методом duplicated():

df.duplicated().sum()

71

Возможные причины появления — ошибка при выгрузке данных

In [52]:
# Удаляем явные дубликаты методом drop_duplicates(), индексы сбросим через reset_index:

df = df.drop_duplicates().reset_index(drop=True)

In [53]:
# Проверим явные дубликаты методом duplicated():

df.duplicated().sum()

0

**Вывод**

Дубликаты устранены, таблица готова к работе

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b>  Удалять дубли можно только после полной обработки данных. Сначала нужно избавиться от неявных дублей - привести значения к нужному решистру, и только потом работать с явными дублями. Попробуй еще раз удалить дубли и посмотри, сколько их осталось

<div class="alert alert-info"> <b>Комментарий студента:</b> Поправил. Сначала — неявные, лишь потом — явные</div>




<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Избавились от всех дублей!

### Лемматизация

In [54]:
df['purpose'].value_counts()

свадьба                                   785
на проведение свадьбы                     759
сыграть свадьбу                           755
операции с недвижимостью                  669
покупка коммерческой недвижимости         655
покупка жилья для сдачи                   647
операции с коммерческой недвижимостью     643
операции с жильем                         641
покупка жилья для семьи                   636
жилье                                     635
покупка жилья                             634
недвижимость                              627
строительство собственной недвижимости    626
операции со своей недвижимостью           623
строительство недвижимости                619
покупка своего жилья                      618
строительство жилой недвижимости          617
покупка недвижимости                      613
ремонт жилью                              602
покупка жилой недвижимости                599
на покупку своего автомобиля              501
заняться высшим образованием      

In [55]:
# В столбце 'purpose' 38 разных значений, мы обнаружили это на этапе знакомства с таблицей.
# Фактически значений значительно меньше. Для того, чтобы выявить реальное количество целей для получения кредита, получим
# леммы из этого списка
# Импортируем библиотеку с функцией лемматизации на русском языке — pymystem3

from pymystem3 import Mystem
m = Mystem()

# создадим функции, в которую будем передавать значения столбца

def create_category_purpose(row):
    lem_purpose = m.lemmatize(row['purpose'])
    if 'автомобиль' in lem_purpose:
        return 'автомобиль'
    if ('жилье' in lem_purpose) or ('недвижимость' in lem_purpose ):
        return 'недвижимость'
    if 'образование' in lem_purpose:
        return 'образование'
    if 'свадьба' in lem_purpose:
        return 'свадьба'

In [56]:
# Применим нашу функцию к столбцу 'purpose':

df['purpose_category'] = df.apply(create_category_purpose, axis=1)

In [57]:
# Посмотрим на таблицу с новым столбцом:

df.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_category  
0               покупка жилья     недвижимость  
1     приобретение автомобиля       автомобиль  
2               покупка жилья     недвижимость  
3  дополнительное образование      образование  
4             сыграть свадьбу          свадьба  
5               покупка жилья     недвижимость  
6           операции с жильем     недвижимость  
7                 образование      образование  
8       на проведение свадьбы          свадьба  
9     покупка жилья для семьи     недвижимость

In [58]:
# Было

df['purpose'].value_counts()

свадьба                                   785
на проведение свадьбы                     759
сыграть свадьбу                           755
операции с недвижимостью                  669
покупка коммерческой недвижимости         655
покупка жилья для сдачи                   647
операции с коммерческой недвижимостью     643
операции с жильем                         641
покупка жилья для семьи                   636
жилье                                     635
покупка жилья                             634
недвижимость                              627
строительство собственной недвижимости    626
операции со своей недвижимостью           623
строительство недвижимости                619
покупка своего жилья                      618
строительство жилой недвижимости          617
покупка недвижимости                      613
ремонт жилью                              602
покупка жилой недвижимости                599
на покупку своего автомобиля              501
заняться высшим образованием      

In [59]:
# Стало 

df['purpose_category'].value_counts()

недвижимость    10704
автомобиль       4258
образование      3970
свадьба          2299
Name: purpose_category, dtype: int64

In [60]:
# Убедимся, что ничего не потеряли

df['purpose'].value_counts().sum()

21231

In [61]:
df['purpose_category'].value_counts().sum()

21231

**Вывод:** **работать с целями получения кредита стало значительно проще**

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Нужные категории найдены </div>

### Категоризация данных

В ключевом столбце `'children'` указано количество детей. В нашем исследовании необходимо узнать, оказывает ли наличие детей влияние на возврат кредита в срок. Разделим всех людей на четыре категории — "один ребенок", "два ребенка", "многодетные" (три и более ребенка) и "без детей". Создадим функцию `'children_qualifier'`, принимающую число и возвращающую значения "с детьми", "многодетные" и "без детей". Далее применим ее к столбцу `'children'` методом `apply`, сохраним категорию в новый столбец `'children_category'` и выведем на экран первые 10 строк таблицы с новым столбцом

In [62]:
def children_qualifier(children):
    if children >= 3:
        return 'многодетные'
    if children == 2:
        return 'два ребенка'
    if children == 1:
        return 'один ребенок'
    return 'без детей'

df['children_category'] = df['children'].apply(children_qualifier)
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_category children_category  
0               покупка жилья     недвижимость      один ребенок  
1     приобретение автомобиля       автомобиль      один ребенок  
2               покупка жилья     недвижимость         без детей  
3  дополнительное образование      образование       многодетные  
4             сыграть свадьбу          свадьба         без детей  
5               покупка жилья     недвижимость         без детей  
6           операции с жильем     недвижимость         без детей  
7                 образование      образование         без детей  
8       на проведение свадьбы          свадьба       два ребенка  
9     покупка жилья для семьи     недвижимость         без детей

В дальнейшем такое упрощение поможет сделать выводы о связи наличия детей (и их количества) и своевременным возвратом кредита

In [63]:
df['children_category'].value_counts()

без детей       14022
один ребенок     4792
два ребенка      2039
многодетные       378
Name: children_category, dtype: int64

В столбце `'total_income'` очень много значений:

In [64]:
df['total_income'].value_counts().sum()

21231

С таким количеством очень неудобно работать. Разобъем данные на категории по методологии Росстата — ведь Росстат не может врать и его методы непогрешимы :)

1. крайняя нищета (доходы ниже прожиточного минимума — до 7-8 тыс.р.)
2. нищета (доходы от одного до двух прожиточных минимума — от 8 до 12 тыс.р.)
3. бедность (доходы от 12 до 20 тысяч рублей в месяц)
4. выше бедности (доходы от 20 до 30 тысяч рублей в месяц)
5. средний достаток — (доходы от 30 до 60 тысяч рублей в месяц)
6. состоятельные — (доходы от 60 до 90 тысяч рублей в месяц)
7. богатые — (доходы от 90 тысяч рублей в месяц)
8. сверхбогатые — (доходы свыше 150 тысяч рублей в месяц)

Сделаем три допущения:
* В `'total_income'` годовой доход, т.е. в функцию будем передавать значение/12 месяцев
* Данные Росстата расчитаны на человека в год, включая детей. Мы посчитаем доход на одного человека, т.к. численность семьи нам неизвестна
* Уменьшим количество категорий. Первые три категории сведем к "бедный", 4 и 5 к "средний класс", 6-8 будут "состоятельный"

Для этого напишем функцию, принимающуя значение дохода и возвращающую название категории. Потом применим через apply эту функцию к столбцу `'total_income'`. Новые данные запишем в столбец `'income_category'` и выведем первые 10 строк таблицы с новым столбцом

In [65]:
def income_qualifier(row):
    if row / 12 < 20000:
        return 'бедный'
    if row / 12 < 60000 and row / 12 >= 20000:
        return 'средний класс'
    if row / 12 >= 60000:
        return 'состоятельный'
    
   
    
df['income_category'] = df['total_income'].apply(income_qualifier)

df.head(10)


children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_category children_category  \
0               покупка жилья     недвижимость      один ребенок   
1     приобретение автомобиля       автомобиль      один ребенок   
2               покупка жилья     недвижимость         без детей   
3  дополнительное образование      образование       многодетные   
4             сыграть свадьбу          свадьба         без детей   
5               покупка жилья     недвижимость         без детей   
6           операции с жильем     недвижимость         без детей   
7                 образование      образование         без детей   
8       на проведение свадьбы          свадьба       два ребенка   
9     покупка жилья для семьи     недвижимость         без детей   

  income_category  
0   средний класс  
1          бедный  
2          бедный  
3   средний класс  
4          бедный  
5   средний класс  
6   средний класс  
7          бедный  
8          бедный  
9          бедный

In [66]:
df['income_category'].value_counts()

бедный           18134
средний класс     3044
состоятельный       53
Name: income_category, dtype: int64

**Вывод**: Мы подготовили ключевой столбец к дальнейшей работе, а людей можно разделить на условные категории по уровню дохода. В нашей таблице, к сожалению, большинство относится к бедному классу.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Категоризация поможет нам ответить на вопросы ниже </div>

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

Чтобы ответить на этот вопрос, сделаем сводную таблицу с индексами = значения столбца `'children_category'`, столбцами будут значения столбца `'debt'`, функция агрегирования — `'count'`(нам нужно посчитать количество людей в сегментах 0 и 1 столбца `'debt'`).
Получив сводную таблицу, добавим к ней столбец `'percent'`, в котором будет выводится отношение просрочивших кредит к общему количеству в процентах

In [67]:
pivot_table_children_category = df.pivot_table(index='children_category', columns='debt', values='dob_years', aggfunc='count')
pivot_table_children_category['percent'] = pivot_table_children_category[1] / (pivot_table_children_category[0] + pivot_table_children_category[1]) * 100
pivot_table_children_category.sort_values(by='percent', ascending=False).head(10)

debt                   0     1   percent
children_category                       
два ребенка         1845   194  9.514468
один ребенок        4351   441  9.202838
многодетные          347    31  8.201058
без детей          12964  1058  7.545286

**Вывод**: как мы видим, заемщики без детей реже уходят в просрочку по кредиту — количество просрочек у "родителей" больше почти на 2%. Антирекордсмены — замещики с двумя детьмя._(ИМХО: крайне удивлен такими цифрами, предпологал, что клиенты с детьми из-за повышенного чувства ответственности, серьезности, зрелости и т.п. будут лучше выполнять обязательства перед банком. Как оказалось, повышенная нагрузка на бюджет сильнее этих высоких моральных качеств)_

<div class="alert alert-warning">
<b>⚠️ Комментарий ревьюера:</b>  Я бы не советовала разделять данные на столь узкие категории. Так мы рискуем упустить скрвтве закономерности

<div class="alert alert-info"> <b>Комментарий студента:</b> Количество категорий увеличил до 4-х.</div>




<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Верно!

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

По аналогии с предыдущим вопросом делаем сводную таблицу, только теперь индексами будут значения столбца `'family_status'`

In [68]:
pivot_table_family_category = df.pivot_table(index='family_status', columns='debt', values='children', aggfunc='count')
pivot_table_family_category['percent'] = pivot_table_family_category[1] / (pivot_table_family_category[0] + pivot_table_family_category[1]) * 100
pivot_table_family_category.sort_values(by='percent', ascending=False).head()

debt                       0    1   percent
family_status                              
Не женат / не замужем   2508  272  9.784173
гражданский брак        3730  383  9.311938
женат / замужем        11290  923  7.557521
в разводе               1095   84  7.124682
вдовец / вдова           884   62  6.553911

**Вывод**: ожидаемо женатые/замужние клиенты реже уходят в просрочку, им же следуют и люди, уже состоявшие в браке (вдовцы/вдовы и разведенные), люди не в браке исполняют обязательства хуже всех. 

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [69]:
pivot_table_income_category = df.pivot_table(index='income_category', columns='debt', values='dob_years', aggfunc='count')
pivot_table_income_category['percent'] = pivot_table_income_category[1] / (pivot_table_income_category[0] + pivot_table_income_category[1]) * 100
pivot_table_income_category.sort_values(by='percent', ascending=False).head(10)

debt                 0     1   percent
income_category                       
бедный           16626  1508  8.315871
состоятельный       49     4  7.547170
средний класс     2832   212  6.964520

**Вывод**: бедные клиенты гасят кредиты хуже всего. В этом разрезе можно углубиться и увеличить сегментирование до 10/20/30 групп, чтобы посмотреть процент просрочек среди среднего класса и состоятельных клиентов

- Как разные цели кредита влияют на его возврат в срок?

In [70]:
pivot_table_purpose_category = df.pivot_table(index='purpose_category', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_purpose_category['percent'] = pivot_table_purpose_category[1] / (pivot_table_purpose_category[0] + pivot_table_purpose_category[1]) * 100
pivot_table_purpose_category.sort_values(by='percent', ascending=False).head(10)

debt                 0    1   percent
purpose_category                     
автомобиль        3861  397  9.323626
образование       3601  369  9.294710
свадьба           2118  181  7.872988
недвижимость      9927  777  7.258969

**Вывод**: ожидаемо лучше всего возвращают ипотечные кредиты (жить на улице не хочет никто), на удивление второе место занимают кредиты на свадьбу. Аутсайдером, на удивление (ведь машина находится в залоге и в случае просрока банк может ее изъять), стали автокредиты — их вовзращают хуже всего. 

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Все так </div>

# Шаг 4. Общий вывод

**Качество входных данных**

Стоит серьезно обратить внимание на качество входных данных. Обратить внимание на:

1. большое (~10%) количество пропусков в `days_employed` и `total_incom`
2. единицу измерения по стажу, дни там или часы?
3. стаж в отрицательных значениях
4. небольшие, менее 1% аномальные значения

По пункту 1, 3 и 4 определить источник(и) проблем

**Цели и размеры кредита**

Уточнить у два момента:
1. Размер получаемых заемных средств. Интересными могу показаться зависимости возврата и абсолютного значения полученных денег, а также связь между возвратом займа и его размера относительно дохода заемщика
2. Вид кредита. Получение потребительского кредита на покупку автомобиля или для покупки квартиры (такое тоже бывает) — возможный вариант. Иногда для избежания залога автомобиля/квартиры именно так и поступают, оформляя потребкредит. Высокий % невозврата по автокредиту дает этой гипотезе право на жизнь

**Результаты исследования**

Худшие заемщики это:

- люди с детьми, хуже всех — люди с двумя детьми
- люди без партнера/живущие в гражданском браке
- люди, берущие деньги на покупку автомобиля
- люди с низким доходом

Кому банкам стоит доверять:
    
- вдовцам и разведенным
- ипотечникам
- людям со средним по размеру доходом
- людям без детей

<div class="alert alert-danger">
<b>❌ Комментарий ревьюера:</b> Вывод - это полный отчет о проделанной работе. Он должен содержать все: от обработки данных до рекомендаций заказчику. Давай расширим его!</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Исправил и дополнил.</div>




<div class="alert alert-info"> <b>Комментарий студента:</b> PS: девушка_с_именем_Саша — это всегда круто!</div>




<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Теперь доволен не только заказчик, но и я:)

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера:</b> Спасибо за работу! Проект получился очень хорошим. Давай сделаем его сильнее:
        
- удалим дубли после полной обработки данных
- поработаем с аномалиями, найдем их, посчитаем их долю и поймем, что с ними делать
- расширим вывод

Оставила направляющие комментарии. Жду твою работу на повторное ревью!
</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьюера V2:</b> Спасибо за проработку комментариев! Теперь все отлично. Если у тебя возникнут вопросы, смело задавай мне их через куратора. Удачи на следующих этапах!